provider/server.go

```diff
@@ -95,37 +101,57 @@ func (server *providerServer) postBlock(w http.ResponseWriter, r *http.Request)
 	}
 	renterID := renterquery[0]
 
-	renter := server.provider.renters[renterID]
-	avail := renter.StorageReserved - renter.StorageUsed
+	renter, exists := server.provider.renters[renterID]
+	if !exists {
+		server.writeResp(w, http.StatusBadRequest,
+			errorResp{Error: "Insufficient space: You have no storage reserved."})
+		return
+	}
+	spaceAvail := renter.StorageReserved - renter.StorageUsed
 
 	// first check block size using the http header
-	if r.ContentLength > avail {
-		msg := fmt.Sprintf("Block of size %d, exceeds available storage %d", r.ContentLength, avail)
-		server.writeResp(w, http.StatusInsufficientStorage, errorResp{Error: msg})
+	if r.ContentLength > spaceAvail {
+		msg := fmt.Sprintf("Block of size %d exceeds available storage %d", r.ContentLength, spaceAvail)
+		server.writeResp(w, http.StatusBadRequest, errorResp{Error: msg})
 		return
 	}
 
+	// create directory for renter's blocks if necessary
+	renterDir := path.Join(server.provider.Homedir, "blocks", renterID)
+	if _, err := os.Stat(renterDir); os.IsNotExist(err) {
+		err := os.MkdirAll(renterDir, 0700)
+		if err != nil {
+			server.logger.Println(err)
+			server.writeResp(w, http.StatusInternalServerError,
+				errorResp{Error: "Unable to save block"})
+			return
+		}
+	}
+
 	// create file
-	path := path.Join(server.provider.Homedir, "blocks", renterID, blockID)
+	path := path.Join(renterDir, blockID)
 	f, err := os.Create(path)
 	if err != nil {
 		server.logger.Println("Unable to create block file. Error: ", err)
-		server.writeResp(w, http.StatusInternalServerError, errorResp{Error: "Unable to save block"})
+		server.writeResp(w, http.StatusInternalServerError,
+			errorResp{Error: "Unable to save block"})
 		return
 	}
 	defer f.Close()
 
 	n, err := io.Copy(f, r.Body)
 	if err != nil {
 		server.logger.Println("Unable to write block. Error: ", err)
-		server.writeResp(w, http.StatusInternalServerError, errorResp{Error: "Unable to save block"})
+		server.writeResp(w, http.StatusInternalServerError,
+			errorResp{Error: "Unable to save block"})
 		return
 	}
 
 	// Verify that received file is the correct size
-	if int64(n) > avail {
+	if int64(n) > spaceAvail {
```

Comment:
Just curious, why are you casting to int64 here?

Reply:
This is Zak's, but I believe the spaceAvail is an int64 since it  is calculated from int64s. You cannot compare an int to an int64 without the explicit cast.